log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java

```diff
@@ -280,16 +274,23 @@ public InputStream getInputStream() {
      */
     public ConfigurationSource resetInputStream() throws IOException {
         if (source != null && data != null) {
-            return new ConfigurationSource(source, data, this.lastModified);
-        } else if (isFile()) {
-            return new ConfigurationSource(new FileInputStream(getFile()), getFile());
-        } else if (isURL() && data != null) {
+            return new ConfigurationSource(source, data, modifiedMillis);
+        }
+        File file = getFile();
+        if (file != null) {
+            return new ConfigurationSource(Files.newInputStream(file.toPath()), getFile());
+        }
+        URL url = getURL();
+        if (url != null && data != null) {
```

Comment:
...the second bug was here: the last modification date propagated correctly from `HttpWatcher` to the `modifiedMillis` mutable field of `ConfigurationSource`. However the code used the immutable `lastModified` field instead.